[#29] Order 사가패턴 적용, Inventory 동시성 문제 해결

dding94 opened the pull request:
Order 도메인 saga - orchestration 패턴 적용
- OrderPayedEvent SelfConsume 비동기 처리

Inventory 재고 감소로직 동시성 이슈
- 조회했을 때의 재고의 값이 변경될 수 있다.
- optimistic lock, pessimistic lock 적절한 방법을 사용해서 적용

dding94 committed:
refactor: 파라미터 이름 변경

dding94 committed:
refactor: 코드 라인 수정

dding94 committed:
feat: orderPayed selfConsume (saga 패턴 - orchestration)

dding94 committed:
test: alreadyOrderCompleted_return_exception 수정

dding94 committed:
refactor: 람다 -> 메서드참조로 변경

dding94 committed:
refactor: 주문에서 비동기작업 시작하여 해당 리스너는 동기로 처리로 변경

dding94 committed:
feat: Pessimistic lock 적용

f-lab-bright commented on order-api/order-presentation/src/main/java/com/flab/order/presentation/request/FailInventoryReducedEvent.java:
```diff
@@ -12,9 +12,13 @@ public final class FailInventoryReducedEvent implements DomainEvent {
 
-    public FailInventoryReducedEvent(Long inventoryId, Long orderId, Integer quantity,
+    public FailInventoryReducedEvent(
+        Long inventoryId,
```
재고가 부족한게 하나가 아닐 수도 있지 않을까요?

dding94 replied:
해당 이벤트는 수정 후 사용되는 곳이 없어 삭제를 고려하고 있습니다..! 만약 사용한다고 하면 재고가 여러개 될 수 있으니 그 점 고려해서 개발하도록 하겠습니다!

f-lab-bright commented on order-api/order-presentation/src/main/java/com/flab/order/presentation/OrderEventHandler.java:
```diff
@@ -24,4 +27,10 @@ public void handle(PaymentCompletedEvent event) {
+
+    @Async
+    @TransactionalEventListener
+    public void orchestrate(OrderPayedEvent event) {
+        orderManager.completedPayed(event);
+    }
```
재고 감소를 시도해보는 로직은 어디로 갔나요??
요 처리기가 주문과 재고를 오케스트레이션하는 부분이라면 재고 감소 요청 후 성공 or 실패 여부에 따라서 다른 처리를 해주도록 로직이 구성되어야 하겠네요.

dding94 replied:
성공 or 실패에 따라 주문의 상태를 변경하도록 수정을 했습니다!
알맞게 구현을 했는지 코드 체크 한번 해주시면 감사하겠습니다.

dding94 committed:
refactor: 주문결제 - 재고감소 flow 리팩토링

롤백상황에서도 재고만 롤백되고 주문은 catch 해서 상태를 cancel 로 변경하도록 수정,
이벤트에 필요없는 내용들 삭제, FailInventory 이벤트는 현재 리스너도 없고 사용되는 곳도 없어 삭제 예정

dding94 committed:
test: 테스트 코드 예외 검증 수정

f-lab-bright commented on inventory-api/inventory-application/src/main/java/com/flab/inventory/application/OrderPayedProcessor.java:
```diff
@@ -32,29 +30,21 @@ public OrderPayedProcessor(
-    @Transactional
-    public void execute(OrderPayedCommand command) {
+    @Transactional(propagation = Propagation.REQUIRES_NEW)
+    public void execute(OrderCompletedCommand command) {
```
inventory인데 OrderPayedProcessor? OrderCompletedCommand? 둘 다 적절하지 않은 컨텍스트 인 것 같네요 :)

f-lab-bright reviewed (commented):
멘토링 시간에 saga pattern 그렸던 것과 구조가 굉장히 다르지 않나요??
이 코드의 흐름이 충분히 이해하기 좋을까요?
inventory가 독립적이려면 orderId만 알지 못하도록 하는게 아니라 주문이라는 컨텍스트 자체가 없어야 하지 않을까요?

dding94 committed:
refactor: 동기 작업을 위한 재고감소 리팩토링

dding94 committed:
refactor: 네이밍 수정 (임시)

dding94 commented on inventory-api/inventory-application/src/main/java/com/flab/inventory/application/DecreaseInventoryProcessor.java:
```diff
@@ -0,0 +1,9 @@
+package com.flab.inventory.application;
+
+import com.flab.inventory.application.command.DecreaseInventoryCommand;
+import com.flab.inventory.domain.ItemQuantity;
+import java.util.List;
+
+public interface DecreaseInventoryProcessor {
+    List<ItemQuantity> execute(DecreaseInventoryCommand command);
+}
```
@f-lab-bright 멘토님 프로세서를 추상화하여 사용한다면 주문에서 해당 프로세서를 동기로 호출 할 수 있을 것 같은데 해당 방법으로 진행을 해도 괜찮을까요?

dding94 committed:
refactor: 주문-재고 동기처리 (진행중)

dding94 commented on order-api/order-domain/src/main/java/com/flab/order/domain/DecreaseInventoryProcessor.java:
```diff
@@ -0,0 +1,8 @@
+package com.flab.order.domain;
+
+import java.util.List;
+
+public interface DecreaseInventoryProcessor {
+
+    List<ItemQuantity> execute(List<ItemQuantity> itemQuantities);
+}
```
@f-lab-bright 제 생각으로는 이 방법이 멘토님이 말씀하신것 처럼 주문모듈에서 구현상세를 주입받아 처리하도록 만들었는데  혹시 이 방법도 방향이 잘못되었을까요?

dding94 committed:
feat: feign client 작업 중 (직렬화 안되는 문제 있음)

dding94 committed:
refactor: feign client 작업 중

- return 값을 받고 성공 /실패 구현해야함.
- order 에서 결과를 받고 실패시 재고는 어떻게 처리해야 하는가

dding94 committed:
refactor: checkStyle

dding94 committed:
docs: ReadMe 수정

dding94 committed:
refactor: 동기 작업을 위한 주문-재고 리팩토링 (진행 중)

- 반환 객체 명확히 지정
- 필요없는 클래스 전부 삭제
- 요청, 응답 로직 수정

dding94 committed:
refactor: 동기 작업을 위한 주문-재고 리팩토링

- 기타 로거 static 붙임
- 도메인 dto 클래스명 수정 -> data 로 통일

dding94 committed:
chore: jpa-open-in-view false 처리

- 트랜잭션 범위 밖으로 나가도 영속성 컨텍스트 유지가 되는 현상 막기 위함, db 커넥션을 계속 유지가 되는 문제

dding94 committed:
refactor: event -> command 로 변환해서 사용하도록 로직 변경

dding94 committed:
test: OrderPayedProcessor 테스트코드 작성

- 재고감소서비스 성공 / 실패 여부에 따른 주문의 상태 테스트

dding94 committed:
refactor: 재고 리팩토링 작업

- decrease 감소 로직 가시성 개선
- dto 접근제어자 수정
- response 객체로 반환하도록 수정 (현재는 재고감소만 나머지 추후 작업 에정)

dding94 committed:
refactor: 불필요한 주문 컨텍스트 삭제

dding94 committed:
test: 테스트코드 수정

- FakeProcessor 가 아닌 실제 비즈니스로직 클래스 테스트 하도록 변경

dding94 committed:
refactor: 접근제어자 수정

2eve2 reviewed (commented):
따로 말씀드린 Inventory내 Order 컨텍스트 제외 부분 빼고 또 고쳐야할 부분을 못찾았습니다.
크리티컬한 지점이 발견된다면 저도 더 공부해보겠습니다.

dding94 committed:
refactor: findByItemIdIn -> findAllById(추가) 로 변경

- 아이템이 아닌 인벤토리 id 로 검색하기 때문

dding94 committed:
refactor: findAllById 추가

- 아이템이 아닌 인벤토리 id 로 검색하기 때문

dding94 committed:
test: Open/Close Processor 테스트코드 작성

f-lab-bright commented on order-api/order-domain/src/main/java/com/flab/order/domain/DecreaseInventoryService.java:
```diff
@@ -0,0 +1,10 @@
+
+public interface DecreaseInventoryService {
+
+    DecreaseInventoryData service(List<ItemQuantityData> itemQuantityDataList);
```
메서드 이름은 좀 더 구체적으로 정하는게 좋겠네요.

f-lab-bright commented on order-api/order-application/src/main/java/com/flab/order/application/OrderPayedProcessor.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    @Transactional
+    public OrderPayedResult execute(OrderPayedCommand command) {
+        var order = orderRepository.findById(command.getOrderId());
+        var data = decreaseInventoryService.service(command.getItemQuantityData());
```
주문이 결제됐을 때 상품 목록을 이벤트로 발행할 수도 있지만, 이 시점에 실제로 감소시켜야 할 상품의 목록을 가져올 수도 있곘네요.

dding94 replied:
서비스의 결과로 재고현황을 가져오기 보다 실제로 감소시켜야 할 상품 목록을 가져왔을 때의 활용도가 command 와 result 를 비교하여 제대로 차감이 됐는지 검증을 위해서 일까요...?

dding94 replied:
@f-lab-bright 멘토님 혹시 이 부분도 조금 더 설명이 가능 하실지..

f-lab-bright reviewed (approved):
LGTM with Comments
항상 코드의 목적이 잘 드러날 수 있도록 네이밍에 신경쓰면 좋을 것 같네요 :)

dding94 committed:
refactor: 메서드 네임 구체적으로 변경

dding94 committed:
refactor: null 반환하지 않도록 수정, isSuccess 제거
